txnsync/encodedgroupsmarshalers.go

```diff
@@ -0,0 +1,818 @@
+	stub.finishDeconstructCompactCertTxnFields()
+}
+
+func (stub *txGroupsEncodingStub) finishDeconstructTxType() {
+	offset := byte(0)
+	count := make(map[int]int)
+	maxcount := 0
+	for _, t := range stub.TxType {
+		count[int(t)]++
+	}
+	for i := range protocol.TxnTypes {
+		if c, ok := count[i]; ok && c > maxcount {
+			offset = byte(i)
+			maxcount = c
+		}
+	}
+	newTxTypes := make([]byte, 0, stub.TotalTransactionsCount)
+	for i := 0; i < int(stub.TotalTransactionsCount); i++ {
+		if stub.TxType[i] != offset {
+			stub.BitmaskTxType.SetBit(i)
+			newTxTypes = append(newTxTypes, stub.TxType[i])
+		}
+	}
+	stub.TxType = newTxTypes
+	stub.TxTypeOffset = offset
+	stub.TxType = compactNibblesArray(stub.TxType)
+	stub.BitmaskTxType.trimBitmask(int(stub.TotalTransactionsCount))
+}
+
+func (stub *txGroupsEncodingStub) deconstructTxnHeader(i int, txn transactions.SignedTxn) {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	if !txn.Txn.Sender.MsgIsZero() {
+		if len(stub.BitmaskSender) == 0 {
+			stub.BitmaskSender = make(bitmask, bitmaskLen)
+			stub.Sender = make([]byte, 0, int(stub.TotalTransactionsCount)*crypto.DigestSize)
+		}
+		stub.BitmaskSender.SetBit(i)
+		stub.Sender = append(stub.Sender, txn.Txn.Sender[:]...)
+	}
+	if !txn.Txn.Fee.MsgIsZero() {
+		if len(stub.BitmaskFee) == 0 {
+			stub.BitmaskFee = make(bitmask, bitmaskLen)
+			stub.Fee = make([]basics.MicroAlgos, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskFee.SetBit(i)
+		stub.Fee = append(stub.Fee, txn.Txn.Fee)
+	}
+	if !txn.Txn.FirstValid.MsgIsZero() {
+		if len(stub.BitmaskFirstValid) == 0 {
+			stub.BitmaskFirstValid = make(bitmask, bitmaskLen)
+			stub.FirstValid = make([]basics.Round, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskFirstValid.SetBit(i)
+		stub.FirstValid = append(stub.FirstValid, txn.Txn.FirstValid)
+	}
+	if !txn.Txn.LastValid.MsgIsZero() {
+		if len(stub.BitmaskLastValid) == 0 {
+			stub.BitmaskLastValid = make(bitmask, bitmaskLen)
+			stub.LastValid = make([]basics.Round, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskLastValid.SetBit(i)
+		stub.LastValid = append(stub.LastValid, txn.Txn.LastValid)
+	}
+	if txn.Txn.Note != nil {
+		if len(stub.BitmaskNote) == 0 {
+			stub.BitmaskNote = make(bitmask, bitmaskLen)
+			stub.Note = make([][]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskNote.SetBit(i)
+		stub.Note = append(stub.Note, txn.Txn.Note)
+	}
+	if txn.Txn.GenesisID != "" {
+		if len(stub.BitmaskGenesisID) == 0 {
+			stub.BitmaskGenesisID = make(bitmask, bitmaskLen)
+		}
+		stub.BitmaskGenesisID.SetBit(i)
+	}
+	if !txn.Txn.GenesisHash.MsgIsZero() {
+		if len(stub.BitmaskGenesisHash) == 0 {
+			stub.BitmaskGenesisHash = make(bitmask, bitmaskLen)
+		}
+		stub.BitmaskGenesisHash.SetBit(i)
+	}
+	if !txn.Txn.Group.MsgIsZero() {
+		if len(stub.BitmaskGroup) == 0 {
+			stub.BitmaskGroup = make(bitmask, bitmaskLen)
+		}
+		stub.BitmaskGroup.SetBit(i)
```

Comment:
I think that we should set the bit only on transactions group of size 1. ( setting this bit on transaction groups > 1 is redundant, since we can extrapolate that information from the size of the transaction group ).